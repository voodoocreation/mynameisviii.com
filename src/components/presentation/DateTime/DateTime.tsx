import cn from "classnames";
import moment from "moment";
import * as React from "react";
import { FormattedRelative, InjectedIntl, injectIntl } from "react-intl";

interface IProps {
  className?: string;
  intl: InjectedIntl;
  isDateOnly?: boolean;
  isRelative?: boolean;
  options?: Intl.DateTimeFormatOptions;
  updateInterval?: number;
  value: string;
}

const DateTime: React.SFC<IProps> = ({
  className,
  intl,
  isDateOnly,
  isRelative,
  options,
  updateInterval,
  value,
  ...props
}) => (
  <time
    className={cn("DateTime", className)}
    dateTime={value}
    title={
      isDateOnly
        ? intl.formatDate(moment(value).toDate(), options)
        : intl.formatTime(moment(value).toDate(), options)
    }
  >
    {isRelative ? (
      <FormattedRelative
        updateInterval={updateInterval}
        value={moment(value).toDate()}
        {...props}
      />
    ) : isDateOnly ? (
      intl.formatDate(moment(value).toDate(), options)
    ) : (
      intl.formatTime(moment(value).toDate(), options)
    )}
  </time>
);

DateTime.defaultProps = {
  isDateOnly: false,
  isRelative: true,
  options: {
    day: "numeric",
    month: "long",
    weekday: "long",
    year: "numeric"
  }
};

export default injectIntl<any>(DateTime);
