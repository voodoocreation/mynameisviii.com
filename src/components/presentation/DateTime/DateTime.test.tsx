import { render } from "enzyme";
import * as React from "react";

import DateTime from "./DateTime";

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    className: "TestDateTime",
    value: "2018-10-20T20:00:00",
    ...fromTestProps
  };

  return {
    actual: fn(<DateTime {...props} />),
    props
  };
};

describe("[presentation] <DateTime />", () => {
  it("renders correctly with minimum props", () => {
    const { actual, props } = setup(render);

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("Saturday, October 20, 2018, 8:00 PM");
    expect(actual.text()).toBe("in 10 months");
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when isDateOnly=true", () => {
    const { actual, props } = setup(render, { isDateOnly: true });

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("Saturday, October 20, 2018");
    expect(actual.text()).toBe("in 10 months");
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when isRelative=false", () => {
    const { actual, props } = setup(render, {
      isRelative: false
    });

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("Saturday, October 20, 2018, 8:00 PM");
    expect(actual.text()).toBe("Saturday, October 20, 2018, 8:00 PM");
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when isRelative=false", () => {
    const { actual, props } = setup(render, {
      isDateOnly: true,
      isRelative: false
    });

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("Saturday, October 20, 2018");
    expect(actual.text()).toBe("Saturday, October 20, 2018");
    expect(actual).toMatchSnapshot();
  });

  it("only renders the time when isRelative=false and options are set", () => {
    const { actual, props } = setup(render, {
      isRelative: false,
      options: {
        hour: "numeric",
        hour12: true,
        minute: "numeric"
      }
    });

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("8:00 PM");
    expect(actual.text()).toBe("8:00 PM");
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when value doesn't include the time ", () => {
    const { actual, props } = setup(render, {
      isDateOnly: true,
      value: "2018-10-20"
    });

    expect(actual.attr("datetime")).toBe(props.value);
    expect(actual.attr("title")).toBe("Saturday, October 20, 2018");
    expect(actual.text()).toBe("in 10 months");
    expect(actual).toMatchSnapshot();
  });
});
