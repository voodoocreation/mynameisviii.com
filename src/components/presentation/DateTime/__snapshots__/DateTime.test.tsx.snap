// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <DateTime /> when isRelative is false and isDateOnly is false matches snapshot 1`] = `
<time
  class="DateTime TestDateTime"
  datetime="2018-10-20T20:00:00"
  title="Saturday, 20 October 2018, 8:00 pm"
>
  Saturday, 20 October 2018, 8:00 pm
</time>
`;

exports[`[presentation] <DateTime /> when isRelative is false and isDateOnly is true matches snapshot 1`] = `
<time
  class="DateTime TestDateTime"
  datetime="2018-10-20"
  title="Saturday, 20 October 2018"
>
  Saturday, 20 October 2018
</time>
`;

exports[`[presentation] <DateTime /> when isRelative is false and only time options are defined matches snapshot 1`] = `
<time
  class="DateTime TestDateTime"
  datetime="2018-10-20T20:00:00"
  title="8:00 pm"
>
  8:00 pm
</time>
`;

exports[`[presentation] <DateTime /> when isRelative is true and isDateOnly is true matches snapshot 1`] = `
<time
  class="DateTime TestDateTime"
  datetime="2018-10-20"
  title="Saturday, 20 October 2018"
>
  in 10 months
</time>
`;
